### PR TITLE
Add missing initialization to CacheProfiler

### PR DESCRIPTION
**Pull request by bmerry**
This fixes using CacheProfiler.register (previously it would only work
if explicitly cleared or used through the context manager protocol).

### CODE REVIEW

File: dask/diagnostics/profile.py [hunk 1]
```diff
@@ -301,6 +301,10 @@ class CacheProfiler(Callback):
     """
 
     def __init__(self, metric=None, metric_name=None):
+        self.results = []
+        self._cache = {}
+        self._dsk = {}
+        self._start_time = None
```
Comment (jcrist) [inline dask/diagnostics/profile.py]: This repeats logic. Why not just call `self.clear()` here?

**Comment by bmerry** [comment]
> This could also use a test to ensure this doesn't break again.

Can you advise on what you think the best approach is? Should I try to write register/unregister-style tests for all the profile classes? Should I make the existing tests parametric so that they run with both styles, or add dedicated tests that just compare the `results` members?

I've also not used pytest much, so I'm not sure how best to do the necessary cleanup to guarantee that unregister gets called if the test fails.

**Comment by jcrist** [comment]
> Can you advise on what you think the best approach is?

This should work fine. I wouldn't worry too much about edge cases of cleanups.

```python
@pytest.mark.parametrize('profiler',
                         [Profiler,
                          lambda: ResourceProfiler(dt=0.01),
                          CacheProfiler])
def test_register(profiler):
    prof = profiler()
    try:
        prof.register()
        get(dsk2, 'c')
        n = len(prof.results)
        assert n > 0
        get(dsk2, 'c')
        assert len(prof.results) > n
    finally:
        prof.unregister()
```

**Comment by bmerry** [comment]
> This should work fine.

Thanks, that seems to work, so I've pushed it. I also added a `skipif("not psutil")` mark. This does mean that the `Profiler` and `CacheProfiler` tests get skipped too if psutil isn't installed, but I'm assuming that it's not worth making the test more complicated to cover that case.

**Comment by bmerry** [comment]
@jcrist that test is failing on Windows. I'm guessing it's because the ResourceProfiler hasn't had time to collect any data? Is it acceptable to just add a sleep to the test?

**Comment by bmerry** [comment]
Actually I think the Windows failure may be due to #2521. For now I've marked it as xfail on win32 - hopefully that mark can be removed once #2521 is fixed.

**Comment by bmerry** [comment]
#2521 was a red herring; it was just a timing issue. I've increased the sleep time and now it's passing.

**Comment by jcrist** [comment]
Yeah, I assume it's due to a slower startup time when creating the tracking process on windows. Could a smaller sleep call still work (perhaps 0.2 seconds)? 1 second per test is much slower than I'd like. Alternatively could change the test to something like:

```python
@pytest.mark.parametrize('profiler',
                         [Profiler,
                          lambda: ResourceProfiler(dt=0.01),
                          CacheProfiler])
def test_register(profiler):
    prof = profiler()
    try:
        prof.register()
        get(dsk2, 'c')
        get(dsk2, 'c')
        assert prof.results
    finally:
        prof.unregister()
```

which would just check that there's no error with multiple scheduler runs, and that results are non-empty after a run.

**Comment by bmerry** [comment]
> Could a smaller sleep call still work (perhaps 0.2 seconds)?

I already tried with 0.3 seconds and that actually failed on one of the Travis runs.

One option would be to adjust things so that only the ResourceProfiler test sleeps (since I presume the Profiler and CacheProfiler tests are deterministic). I'll also see if it's practical to replace the sleep with something that polls until results appear (with a longish timeout) so that the test can run quickly.

**Comment by bmerry** [comment]
I've fixed the test by changing the ResourceProfiler to always collect at least one sample before returning data. I'm not 100% sure that won't cause issues if there are many very short start/stop cycles, in which case the ResourceProfiler could become a bottleneck.

**Comment by jcrist** [comment]
Looks good to me. Thanks @bmerry!

**Pull request merged**